### DISCUSSION TITLE
Trying to implement Heterogeneous graph but training fails.

### DISCUSSION THREAD

**Discussion by goodrahstar** [pyg-team/pytorch_geometric#4148] [General]
Hi Team, 
I am trying to train a heterogeneous graph but i am getting errors when trying to train it.
The heterograph looks like this
```
HeteroData(
  users={
    x=[9000, 1550],
    y=[9000]
  },
  engagement={ x=[9000, 9] },
  product={ x=[18, 18] },
  (users, shows, engagement)={ edge_index=[2, 9000] },
  (users, purchase , product)={ edge_index=[2, 18000] }
)
```

And the training architecture looks like this
```
Sequential(
  (0): SAGEConv((-1, -1), 64)
  (1): ReLU(inplace=True)
  (2): SAGEConv((-1, -1), 64)
  (3): ReLU(inplace=True)
  (4): SAGEConv((-1, -1), 64)
  (5): ReLU(inplace=True)
  (6): Linear(-1, 2, bias=True)
)
```

Here is the code for the model architecture.
```
num_classes=2
model = Sequential('x, edge_index', [
    (SAGEConv((-1, -1), 64), 'x, edge_index -> x'),
    ReLU(inplace=True),
    (SAGEConv((-1, -1), 64), 'x, edge_index -> x'),
    ReLU(inplace=True),
    (SAGEConv((-1, -1), 64), 'x, edge_index -> x'),
    ReLU(inplace=True),    
    (Linear(-1, num_classes), 'x -> x'),
])
model = to_hetero(model, data.metadata(), aggr='sum').to(device)
```
When trying to train the model, I get the following error.

Execution: ```model(data.x_dict, data.edge_index_dict)```
Error:  ```RuntimeError: index 9018 is out of bounds for dimension 0 with size 9000```

 How do i make sure the dimensions are correct and overcome this issue.

Thanks

**Comment by rusty1s** [comment]
For heterogeneous graphs, you need to ensure that the following runs through:
```python
for edge_type in data.edge_types:
    src, _, dst = edge_type
    assert data[edge_type].edge_index[0].max() < data[src].num_nodes
    assert data[edge_type].edge_index[1].max() < data[dst].num_nodes
```
Otherwise, you may want to revisit your data processing pipeline to ensure that all edges points to valid node indices. You can use [this tutorial](https://pytorch-geometric.readthedocs.io/en/latest/notes/load_csv.html) as a reference.

**Reply by goodrahstar** [reply]
Thanks for the response @rusty1s , This was a great tip.
